src/core/task_mngr.c

```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+void run_systemtasks_preexecute(void){
+	for(int i = 0; i < TASK_SYSTEM_PREEXECUTE_LIST_SIZE; i++){
+		if(task_system_preexecute[i] != NULL){
+			task_system_preexecute[i]->run(task_system_preexecute[i]->data);
+		}
+	}
+}
+
+
+void run_systemtasks_postexecute(void){
+	for(int i = 0; i < TASK_SYSTEM_POSTEXECUTE_LIST_SIZE; i++){
+		if(task_system_postexecute[i] != NULL){
+			task_system_postexecute[i]->run(task_system_postexecute[i]->data);
+		}
+	}
+}
+
+void run_calculate_priority(void){
+	for(int i = 0; i < TASK_USER_LIST_SIZE; i++){
+		if(task_user_list[i] != NULL){
+			if(task_user_list[i]->data->state != TASK_DONE)
+				task_user_list[i]->calculate_priority(task_user_list[i]->data);
+		}
+	}
+}
+
+/**
+ * @brief sets _task to active task
+ * @param _task to be set as active
+ */
+void set_start_task(task_t *_task){
+	task_user_active = _task;
+}
+
+/**
+ * @brief adds task to corresponding lists of tasks
+ * @param _task pointer to task that should be added
+ */
+void add_task(task_t *_task){
+
+
+	if(_task->type == TASK_USER){
+		if(task_user_couter < TASK_USER_LIST_SIZE){
+			task_user_list[task_user_couter++] = _task;
+		}else{
+			error("TASK_USER_LIST is to small, skipping task adding");
+		}
+	}else if(_task->type == TASK_SYSTEM_PREEXECUTE){
+		if(task_system_preexecute_couter < TASK_SYSTEM_PREEXECUTE_LIST_SIZE){
+			task_system_preexecute[task_system_preexecute_couter++] = _task;
+		}else{
+			error("TASK_SYSTEM_PREEXECUTE_LIST is to small, skipping task adding");
+		}
+	}else if(_task->type == TASK_SYSTEM_POSTEXECUTE){
+		if(task_system_postexecute_couter < TASK_SYSTEM_POSTEXECUTE_LIST_SIZE){
+			task_system_postexecute[task_system_postexecute_couter++] = _task;
+		}else{
+			error("TASK_SYSTEM_POSTEXECUTE_LIST is to small, skipping task adding");
+		}
+	}else if(_task->type == TASK_BACKPROC){
+		if(task_user_backproc_couter < TASK_USER_BACKPROC_LIST_SIZE){
+			task_user_backproc[task_user_backproc_couter++] = _task;
+		}else{
+			error("TASK_USER_BACKPROC_LIST is to small, skipping task adding");
+		}
+	}else{
+		error("Task manager: adding task: unknown task type");
+	}
+
+}
+
+/*
+ * @brief finds ready task (TASK_READY status) with biggest priority.
+ * @param _task pointer to task
+ */
+task_t* find_next_task(void){
+	task_t *_task = NULL;
+	int state = 0;
+
+	// update priority and status of tasks
+	run_calculate_priority();
+
+	// find task with biggest priority
+	for(int i = 0; i < task_user_couter; i++){
+		switch (state) {
+		case 0:{
+			if(task_user_list[i]->data->state == TASK_READY){
+				_task = task_user_list[i];
+				state++;
+			}
+		}break;
+
+		default:{
+			if(_task->data->priority < task_user_list[i]->data->priority &&
+					task_user_list[i]->data->state == TASK_READY){
+				_task = task_user_list[i];
+			}
+		}break;
+		} // end switch
+	}
+	if(_task != NULL) {
+		info("Starting task: %s, by priority: %d",_task->task_name, _task->data->priority);
+	}else{
+		error("No Ready task in list ! exiting");
```

Comment:
Actually no, that task is part of FRTOS (github) task. This is just task manager.